OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnDeploy.cs

```diff
@@ -131,52 +131,56 @@ public IEnumerable<IOrderTargeter> Orders
 		public Order IssueOrder(Actor self, IOrderTargeter order, Target target, bool queued)
 		{
 			if (order.OrderID == "GrantConditionOnDeploy")
-			{
-				var gcodorder = new Order(order.OrderID, self, queued);
-
-				// HACK HACK HACK
-				if (Info.SynchronizeDeployment)
-				{
-					var actors = self.World.Selection.Actors.Select(x => x.ActorID.ToString());
-					gcodorder.TargetString = string.Join(",", actors);
-				}
-
-				return gcodorder;
-			}
+				return new Order(order.OrderID, self, queued);
 
 			return null;
 		}
 
 		Order IIssueDeployOrder.IssueDeployOrder(Actor self, bool queued)
 		{
-			var gcodorder = new Order("GrantConditionOnDeploy", self, queued);
-			if (Info.SynchronizeDeployment)
-			{
-				var actors = self.World.Selection.Actors.Select(x => x.ActorID.ToString());
-				gcodorder.TargetString = string.Join(",", actors);
-			}
-
-			return gcodorder;
+			return new Order("GrantConditionOnDeploy", self, queued);
 		}
 
-		bool IIssueDeployOrder.CanIssueDeployOrder(Actor self) { return !IsTraitPaused && !IsTraitDisabled; }
+		bool IIssueDeployOrder.CanIssueDeployOrder(Actor self) {
+			if (!IsTraitPaused && !IsTraitDisabled && IsGroupDeployNeeded(self))
+				return true;
+			else
+				return false;
+		}
 
-		bool IsGroupDeployNeeded(Actor self, string actorString)
+		bool IsGroupDeployNeeded(Actor self)
 		{
-			if (string.IsNullOrEmpty(actorString))
-				return false;
+			var actors = self.World.Selection.Actors;
 
-			var actorIDs = actorString.Split(',').Select(x => { uint result; uint.TryParse(x, out result); return result; });
-			var actors = self.World.Actors.Where(x => x.IsInWorld && !x.IsDead && actorIDs.Contains(x.ActorID));
+			if (!Info.SynchronizeDeployment)
+				return true;
+
+			int deployedCount = 0;
+			int undeployedCount = 0;
 
 			foreach (var a in actors)
 			{
-				var gcod = a.TraitOrDefault<GrantConditionOnDeploy>();
-				if (gcod != null && gcod.DeployState != DeployState.Deployed)
+				GrantConditionOnDeploy gcod = null;
+				if (!a.IsDead && a.IsInWorld)
+					gcod = a.TraitOrDefault<GrantConditionOnDeploy>();
+
+				if (gcod != null && (gcod.DeployState == DeployState.Deploying || gcod.DeployState == DeployState.Deployed))
+					deployedCount += 1;
+
+				if (gcod != null && (gcod.DeployState == DeployState.Undeploying || gcod.DeployState == DeployState.Undeployed))
+					undeployedCount += 1;
+			}
+
+			if (deployedCount > 0 && undeployedCount > 0)
```

Comment:
Wouldn't it make sense to put this evaulation within the foreach block to an earlier return for performance? I mean, as soon as this check becomes true, you no longer need to evaluate the actors above anymore.